azuread/resource_domain_service.go

```diff
@@ -0,0 +1,492 @@
+	filteredSync := d.Get("filtered_sync").(string)
+	ldapsSettings := d.Get("ldaps_settings").([]interface{})
+	notificationSettings := d.Get("notification_settings").([]interface{})
+	subnetId := d.Get("subnet_id").(string)
+
+	domainService := aad.DomainService{
+		Location: &location,
+		DomainServiceProperties: &aad.DomainServiceProperties{
+			DomainName:             &name,
+			DomainSecuritySettings: expandArmDomainServiceDomainSecuritySettings(domainSecuritySettings),
+			FilteredSync:           aad.FilteredSync(filteredSync),
+			LdapsSettings:          expandArmDomainServiceLdapsSettings(ldapsSettings),
+			NotificationSettings:   expandArmDomainServiceNotificationSettings(notificationSettings),
+			SubnetID:               &subnetId,
+		},
+	}
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, name, domainService)
+	if err != nil {
+		return fmt.Errorf("Error creating Domain Service %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for creation of Domain Service %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	// Domain Services is 2 controllers running in the cloud
+	// the create api completes Once the first domain controller is up and running, 1 ip address will show up
+	// Afterwards, there will be an additional domain controller creating, then the ip address will update and become 2
+	// the Azure Portal blocks users from modifying the domain service until both domain controllers are up
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"pending"},
+		Target:       []string{"available"},
+		Refresh:      domainServiceControllerRefreshFunc(ctx, &client, resourceGroup, name),
+		Delay:        30 * time.Second,
+		PollInterval: 10 * time.Second,
+		Timeout:      1 * time.Hour,
+	}
+
+	if domainControllerIPAddress, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for both Domain Service Controller up, err: %+v", err)
+	} else {
+		// Update DNS server settings of the virtual network
+		parsedId, err := azure.ParseAzureResourceID(subnetId)
+		if err != nil {
+			return err
+		}
+		resourceGroupName := parsedId.ResourceGroup
+		virtualNetworkName := parsedId.Path["virtualNetworks"]
+
+		resp, err := vnetClient.Get(ctx, resourceGroupName, virtualNetworkName, "")
+		if err != nil {
+			return fmt.Errorf("Error readding Virtual Network %q (Resource Group %q): %+v", virtualNetworkName, resourceGroupName, err)
+		}
+		dns := domainControllerIPAddress.([]string)
+		resp.DhcpOptions = &network.DhcpOptions{
+			DNSServers: &dns,
+		}
+		if _, err := vnetClient.CreateOrUpdate(ctx, resourceGroupName, virtualNetworkName, resp); err != nil {
+			return fmt.Errorf("Error updating DNS server to %+v for Virtual Network %q (Resource Group %q): %+v", domainControllerIPAddress, virtualNetworkName, resourceGroupName, err)
+		}
```

Comment:
This should be done separately from this resource in the AzuerRM provider. Most likely as a `azurerm_vnet_dhcp` options resource given this depends on a subnet which depends on a vnet